src/primitives/line/dotted_bresenham.rs

```diff
@@ -0,0 +1,246 @@
+//! A variant of the bresenham algorithm that is used to select
+//! points from a bresenham line in order to draw a dotted line.
+
+use super::{bresenham::MajorMinor, Line, Points};
+use crate::geometry::Point;
+use core::cmp::{max, min};
+
+/// Dotted bresenham parameters.
+///
+/// [`super::bresenham::BresenhamParameters`] describes a major and
+/// a minor step, classically vectors along opposing axes.
+///
+/// [`DottedBresenhamParameters`] describes a scalar major step and a
+/// scalar minor step. It is used to draw a dotted line with the
+/// desired number of dots by selecting the appropriate points in a
+/// bresenham line.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[cfg_attr(feature = "defmt", derive(::defmt::Format))]
+pub struct DottedBresenhamParameters {
+    /// Error threshold.
+    ///
+    /// If the accumulated error exceeds the threshold a minor move is made.
+    pub error_threshold: i32,
+
+    /// Change in error for major and minor steps.
+    pub error_step: MajorMinor<i32>,
+
+    /// Change in index for major and minor steps.
+    pub index_step: MajorMinor<usize>,
+}
+
+impl DottedBresenhamParameters {
+    /// Create a new bresenham parameters object.
+    pub fn new(line: &Line, nb_dots_desired: i32) -> Self {
+        let delta = line.delta().abs();
+        let line_max_index = delta.x.max(delta.y);
+
+        // Enforce the use of at most `line_max_index + 1` dots.
+        let mut nb_dots = min(nb_dots_desired, line_max_index + 1);
+        // Then enforce the use of at least 2 dots to prevent division by 0 when the line is reduced to a point.
+        nb_dots = max(2, nb_dots);
+
+        let integer_quotient = line_max_index / (nb_dots - 1);
+        let remainder = line_max_index - integer_quotient * (nb_dots - 1);
+
+        let index_step = MajorMinor::new(integer_quotient as usize, 1);
+        let error_threshold = nb_dots - 1;
+
+        Self {
+            error_threshold,
+            error_step: MajorMinor::new(remainder * 2, error_threshold * 2),
+            index_step,
+        }
+    }
+
+    /// Increases the error by a major step.
+    ///
+    /// If the error threshold is reached the error is reduced by a minor step and
+    /// `true` is returned.
+    pub fn increase_error(&self, error: &mut i32) -> bool {
+        *error += self.error_step.major;
+        if *error >= self.error_threshold {
+            *error -= self.error_step.minor;
+
+            true
+        } else {
+            false
+        }
+    }
```

Comment:
That's some not so great code I wrote 5 years ago and using that confusing `&mut i32` parameter isn't necessary here.

I don't think there is a reason to keep the split between `DottedBresenhamParameters` and `DottedBresenham`. I would suggest to merge them, or, if you prefer, even merge everything into `DottedLinePoints`.